File: necklace-animation/process_images.py
import numpy as np
from PIL import Image
import os, shutil

PNG_PATH = r'C:\Users\ziyad\Downloads\2c59c371-2fa1-41e0-8533-1d73ae9277df.png'
GIF_PATH = r'C:\Users\ziyad\Downloads\54c00e71c0e4c2f5034b3e1f4a46fe0e.gif'
OUT = r'C:\Users\ziyad\.codely\Default\necklace-animation\assets'

os.makedirs(OUT, exist_ok=True)
shutil.copy2(GIF_PATH, os.path.join(OUT, 'background.gif'))

img = np.array(Image.open(PNG_PATH).convert('RGB'), dtype=np.float32)
H, W, _ = img.shape
brightness = img.mean(axis=2)
print(f"Image: {W}x{H}")

# --- Pass 1: Percentile-based background estimation ---
small_h, small_w = 8, 12
block_h = H // small_h
block_w = W // small_w

small = np.zeros((small_h, small_w, 3), dtype=np.float32)
for sy in range(small_h):
    for sx in range(small_w):
        y0, y1 = sy * block_h, min((sy + 1) * block_h, H)
        x0, x1 = sx * block_w, min((sx + 1) * block_w, W)
        small[sy, sx, :] = np.percentile(img[y0:y1, x0:x1, :], 10, axis=(0, 1))

bg_pil = Image.fromarray(small.clip(0, 255).astype(np.uint8), 'RGB').resize((W, H), Image.BILINEAR)
bg = np.array(bg_pil, dtype=np.float32)
diff = np.sqrt(np.sum((img - bg) ** 2, axis=2))

THRESH = 38
GRAD = 12
alpha = np.full((H, W), 255.0, dtype=np.float32)
alpha[diff < THRESH] = 0
grad_mask = (diff >= THRESH) & (diff < THRESH + GRAD)
alpha[grad_mask] = ((diff[grad_mask] - THRESH) / GRAD) * 255
# Force bright metallic pixels opaque
alpha[brightness > 115] = 255

# --- Pass 2: Refinement - limited flood fill from boundary ---
# For opaque pixels adjacent to transparent ones, check similarity to transparent neighbors
REFINE_TOL = 22
for iteration in range(8):
    transp = alpha < 128
    
    # Find opaque pixels adjacent to transparent
    adj = np.zeros_like(transp)
    adj[1:, :] |= transp[:-1, :]
    adj[:-1, :] |= transp[1:, :]
    adj[:, 1:] |= transp[:, :-1]
    adj[:, :-1] |= transp[:, 1:]
    
    boundary = adj & ~transp
    if not boundary.any():
        print(f"Refinement converged at iteration {iteration}")
        break
    
    # Average color of transparent neighbors (3x3)
    padded_t = np.pad(transp, ((1,1),(1,1)), mode='constant', constant_values=False)
    t_sum = np.zeros_like(img)
    t_count = np.zeros((H, W), dtype=np.float32)
    for dy in range(3):
        for dx in range(3):
            m = padded_t[dy:dy+H, dx:dx+W]
            t_sum += img * m[:, :, None]
            t_count += m
    t_count[t_count == 0] = 1
    t_avg = t_sum / t_count[:, :, None]
    
    pixel_diff = np.sqrt(np.sum((img - t_avg) ** 2, axis=2))
    # Only remove dark pixels (avoid removing necklace)
    new_transp = boundary & (pixel_diff < REFINE_TOL) & (brightness < 110)
    
    if not new_transp.any():
        print(f"Refinement converged at iteration {iteration}")
        break
    
    # Apply with gradient
    new_alpha = np.where(new_transp, np.minimum(alpha, (1.0 - (pixel_diff / REFINE_TOL)) * 255), alpha)
    alpha = new_alpha

alpha_u8 = alpha.clip(0, 255).astype(np.uint8)
transp_pct = (alpha_u8 == 0).sum() / (H * W) * 100
opaque_pct = (alpha_u8 == 255).sum() / (H * W) * 100
semi_pct = ((alpha_u8 > 0) & (alpha_u8 < 255)).sum() / (H * W) * 100
print(f"Transparent: {transp_pct:.1f}%, Opaque: {opaque_pct:.1f}%, Semi: {semi_pct:.1f}%")

# Save RGBA
rgba = np.dstack([img.astype(np.uint8), alpha_u8])
Image.fromarray(rgba, 'RGBA').save(os.path.join(OUT, 'necklace-full.png'))

# Split into 3 parts
third = W // 3
Image.fromarray(rgba[:, :third, :], 'RGBA').save(os.path.join(OUT, 'necklace-left.png'))
Image.fromarray(rgba[:, third:2*third, :], 'RGBA').save(os.path.join(OUT, 'necklace-center.png'))
Image.fromarray(rgba[:, 2*third:, :], 'RGBA').save(os.path.join(OUT, 'necklace-right.png'))

# Verify
verify = Image.open(os.path.join(OUT, 'necklace-full.png'))
print(f"Saved mode: {verify.mode}, size: {verify.size}")
print("Done!")

File: necklace-animation/process-images.mjs
import sharp from 'sharp';
import { copyFileSync, mkdirSync } from 'fs';

const PNG_PATH = 'C:\\Users\\ziyad\\Downloads\\2c59c371-2fa1-41e0-8533-1d73ae9277df.png';
const GIF_PATH = 'C:\\Users\\ziyad\\Downloads\\54c00e71c0e4c2f5034b3e1f4a46fe0e.gif';
const OUT = 'C:\\Users\\ziyad\\.codely\\Default\\necklace-animation\\assets';

mkdirSync(OUT, { recursive: true });
copyFileSync(GIF_PATH, `${OUT}/background.gif`);

const { data, info } = await sharp(PNG_PATH).removeAlpha().raw().toBuffer({ resolveWithObject: true });
const { width, height, channels } = info;

console.log(`Image: ${width}x${height}`);

// For each row, estimate the background color by sampling pixels from the
// left and right edges. Since the background is a gradient, we sample
// multiple pixels and take the darkest (background is darker than the chain).
const N_SAMPLES = 10;
const EDGE_OFFSET = 2;
const bgLeft = new Array(height);
const bgRight = new Array(height);

for (let y = 0; y < height; y++) {
  // Left edge: sample N pixels, keep darkest 3
  const leftSamples = [];
  for (let s = 0; s < N_SAMPLES; s++) {
    const idx = (y * width + EDGE_OFFSET + s) * channels;
    leftSamples.push({
      r: data[idx], g: data[idx + 1], b: data[idx + 2],
      bright: (data[idx] + data[idx + 1] + data[idx + 2]) / 3
    });
  }
  leftSamples.sort((a, b) => a.bright - b.bright);
  const ld = leftSamples.slice(0, 3);
  bgLeft[y] = {
    r: ld.reduce((s, p) => s + p.r, 0) / 3,
    g: ld.reduce((s, p) => s + p.g, 0) / 3,
    b: ld.reduce((s, p) => s + p.b, 0) / 3
  };

  // Right edge: sample N pixels, keep darkest 3
  const rightSamples = [];
  for (let s = 0; s < N_SAMPLES; s++) {
    const idx = (y * width + (width - 1 - EDGE_OFFSET - s)) * channels;
    rightSamples.push({
      r: data[idx], g: data[idx + 1], b: data[idx + 2],
      bright: (data[idx] + data[idx + 1] + data[idx + 2]) / 3
    });
  }
  rightSamples.sort((a, b) => a.bright - b.bright);
  const rd = rightSamples.slice(0, 3);
  bgRight[y] = {
    r: rd.reduce((s, p) => s + p.r, 0) / 3,
    g: rd.reduce((s, p) => s + p.g, 0) / 3,
    b: rd.reduce((s, p) => s + p.b, 0) / 3
  };
}

// Smooth background estimates (moving average over 7 rows)
function smoothRow(arr, radius) {
  const out = new Array(arr.length);
  for (let y = 0; y < arr.length; y++) {
    let r = 0, g = 0, b = 0, n = 0;
    for (let dy = -radius; dy <= radius; dy++) {
      const yy = Math.max(0, Math.min(arr.length - 1, y + dy));
      r += arr[yy].r; g += arr[yy].g; b += arr[yy].b; n++;
    }
    out[y] = { r: r / n, g: g / n, b: b / n };
  }
  return out;
}

const smoothL = smoothRow(bgLeft, 4);
const smoothR = smoothRow(bgRight, 4);

// Build RGBA: pixel is transparent if close to estimated background
const rgba = Buffer.alloc(width * height * 4);
const THRESHOLD = 22;
const GRADIENT = 23;
let transparentCount = 0;

for (let y = 0; y < height; y++) {
  for (let x = 0; x < width; x++) {
    const t = x / (width - 1);
    // Interpolate between left and right background colors
    const bgR = smoothL[y].r * (1 - t) + smoothR[y].r * t;
    const bgG = smoothL[y].g * (1 - t) + smoothR[y].g * t;
    const bgB = smoothL[y].b * (1 - t) + smoothR[y].b * t;

    const i = (y * width + x) * channels;
    const r = data[i], g = data[i + 1], b = data[i + 2];

    const dist = Math.sqrt((r - bgR) ** 2 + (g - bgG) ** 2 + (b - bgB) ** 2);

    let alpha;
    if (dist < THRESHOLD) {
      alpha = 0;
      transparentCount++;
    } else if (dist < THRESHOLD + GRADIENT) {
      alpha = Math.round(((dist - THRESHOLD) / GRADIENT) * 255);
    } else {
      alpha = 255;
    }

    const o = (y * width + x) * 4;
    rgba[o] = r;
    rgba[o + 1] = g;
    rgba[o + 2] = b;
    rgba[o + 3] = alpha;
  }
}

console.log(`Transparent: ${transparentCount} pixels (${(transparentCount / (width * height) * 100).toFixed(1)}%)`);

// Save full necklace
await sharp(rgba, { raw: { width, height, channels: 4 } })
  .png()
  .toFile(`${OUT}/necklace-full.png`);

// Split into 3 parts
const third = Math.floor(width / 3);

await sharp(rgba, { raw: { width, height, channels: 4 } })
  .extract({ left: 0, top: 0, width: third, height })
  .png()
  .toFile(`${OUT}/necklace-left.png`);

await sharp(rgba, { raw: { width, height, channels: 4 } })
  .extract({ left: third, top: 0, width: third, height })
  .png()
  .toFile(`${OUT}/necklace-center.png`);

await sharp(rgba, { raw: { width, height, channels: 4 } })
  .extract({ left: third * 2, top: 0, width: width - third * 2, height })
  .png()
  .toFile(`${OUT}/necklace-right.png`);

console.log('Done! All images saved.');
